packages/publish/src/publish.ts

```diff
@@ -0,0 +1,210 @@
+/**
+ * Copyright 2019 Google LLC. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// define types for a few modules used by probot that do not have their
+// own definitions published. Before taking this step, folks should first
+// check whether type bindings are already published.
+
+import { Application, Context } from 'probot';
+import { execSync } from 'child_process';
+import fetch from 'node-fetch';
+import * as tar from 'tar';
+import * as uuid from 'uuid';
+import { promises as fs } from 'fs';
+import { resolve } from 'path';
+
+const CONFIGURATION_FILE_PATH = 'publish.yml';
+
+interface Configuration {
+  project?: string;
+  secretId?: string;
+}
+
+interface PublishConfig {
+  token: string;
+  registry: string;
+}
+
+interface Secret {
+  payload: { [key: string]: Buffer };
+}
+
+import { SecretManagerServiceClient } from '@google-cloud/secret-manager';
+const sms = new SecretManagerServiceClient();
+
+function handler(app: Application) {
+  app.on('release.released', async context => {
+    const repoName = context.payload.repository.name;
+    const remoteConfiguration: Configuration | null = (await context.config(
+      CONFIGURATION_FILE_PATH
+    )) as Configuration | null;
+
+    // Skip publication unless it's explicitly enabled.
+    if (!remoteConfiguration) {
+      app.log.info(`publish not configured for (${repoName})`);
+      return;
+    }
+
+    if (context.payload.release.tarball_url) {
+      // Create a temporary directory to unpack release tarball to:
+      const unpackPath = handler.unpackPath();
+      app.log.info(`creating tmp directory ${unpackPath}`);
+      await fs.mkdir(unpackPath, {
+        recursive: true,
+      });
+
+      // Unpack the tarball to the directory we just created:
+      await fetch(context.payload.release.tarball_url).then(res => {
```

Comment:
:+1: yeah, I figured the extenuating circumstance of using `tar`, which isn't particularly suited to a promise API, made it an acceptable evil ... I wanted to use this library since it's the one used by `npm`, and I trust it.